apps/slack/slack_listeners.py

```diff
@@ -0,0 +1,118 @@
+"""
+This module contains methods for handling event payloads from subscribed
+events.
+
+Manage event subscriptions at:
+
+    https://api.slack.com/apps/<APP ID>/event-subscriptions
+"""
+
+import logging
+import re
+
+from django.db.models import Q
+from slack_bolt import BoltContext, BoltResponse
+
+from apps.channels.const import SLACK_ALL_CHANNELS
+from apps.channels.datamodels import SlackMessage
+from apps.channels.models import ChannelPlatform, ExperimentChannel
+from apps.chat.channels import SlackChannel
+from apps.experiments.models import ExperimentSession
+from apps.slack.models import SlackInstallation
+from apps.slack.slack_app import app
+from apps.slack.utils import make_session_external_id
+
+logger = logging.getLogger("slack.events")
+
+
+def register_listeners():
+    """Register these after DB setup is complete to avoid hitting the
+    wrong DB e.g. during tests"""
+    app.use(load_installation_and_team)
+    app.event({"type": "message"})(new_message)
+
+
+def new_message(event, context: BoltContext):
+    thread_ts = event.get("thread_ts", None)
+    channel_id = event.get("channel")
+
+    session = None
+    if thread_ts:
+        session = get_session_for_thread(context["team"], channel_id, thread_ts)
+
+    is_bot_mention = context.bot_user_id in event.get("text", "")
+    if is_bot_mention or session:
+        respond_to_message(event, context, session)
+
+
+def respond_to_message(event, context: BoltContext, session=None):
+    context.ack("...")
+
+    channel_id = event.get("channel")
+    thread_ts = event.get("thread_ts", None) or event["ts"]
+    experiment_channel = get_experiment_channel(context["team"], channel_id)
+    if not experiment_channel:
+        context.say("There are no bots associated with this channel.", thread_ts=thread_ts)
+        return
+
+    slack_user = event.get("user")
+
+    if not session:
+        external_id = make_session_external_id(channel_id, thread_ts)
+        session = SlackChannel.start_new_session(
+            experiment_channel.experiment, experiment_channel, slack_user, external_id=external_id
+        )
+
+    # strip out the mention
+    message_text = re.sub(rf"<@?{context.bot_user_id}>", "", event["text"])
+    message = SlackMessage(channel_id=channel_id, thread_ts=thread_ts, message_text=message_text)
+
+    ocs_channel = SlackChannel(experiment_channel, session, send_response_to_user=False)
+    response = ocs_channel.new_user_message(message)
+    context.say(response, thread_ts=thread_ts)
```

Comment:
In the `send_text_to_user` method in `SlackChannel`, we call `self.messaging_service.send_text_message(...)`. Would `context.say(response, thread_ts=thread_ts)` not cause a duplicate message to be send? (I think it wouldn't, otherwise you would have picked it up, but this line confuses me)

Reply:
I chose to do it this way rather than using the messaging service because there is already an authed client available here.

The key is in the `send_response_to_user` param of the `SlackChannel` class which I set to `False` here to prevent it sending the message. When the channel is used for adhoc bot messages then `send_response_to_user` would be `True` and the message would be sent via the messaging service.

Reply:
Right! I missed the `send_response_to_user=False` :man_facepalming:. Maybe add a comment explaining the decision?

Reply:
I'm still to read the slack docs on how this integration works exactly, so maybe I'll find the answer in there, but there's two ways we "send a message". 1 is here by doing `context.say(..)`, the other is using the `WebClient`. Why is this necessary?

Reply:
`context.say` is just a shortcut so that you don't have to create a new web client and authenticate it. It's only available when processing an incoming event and it uses `WebClient` under the hood.

If you want to send messages outside of event handling you have to construct a new WebClient